src/index.ts

```diff
@@ -0,0 +1,241 @@
+import {
+  params,
+  fnv,
+  fnvBuffer,
+  bufReverse,
+  getEpoc,
+  getCacheSize,
+  getFullSize,
+  getSeed
+} from './util'
+import type { LevelUp } from 'levelup'
+const async = require('async')
+const xor = require('buffer-xor')
+
+// ethereumjs-block v3 will have types
+type Block = any
+type BlockHeader = any
+
+export default class Ethash {
+  dbOpts: Object
+  cacheDB?: LevelUp
+  cache: Buffer[]
+  epoc?: number
+  fullSize?: number
+  cacheSize?: number
+  seed?: Buffer
+
+  constructor(cacheDB?: LevelUp) {
+    this.dbOpts = {
+      valueEncoding: 'json'
+    }
+    this.cacheDB = cacheDB
+    this.cache = []
+  }
+
+  mkcache(cacheSize: number, seed: Buffer) {
+    // console.log('generating cache')
+    // console.log('size: ' + cacheSize)
+    // console.log('seed: ' + seed.toString('hex'))
+    const n = Math.floor(cacheSize / params.HASH_BYTES)
+    const o = [keccak(seed, 512)]
+
+    let i
+    for (i = 1; i < n; i++) {
+      o.push(keccak(o[o.length - 1], 512))
+    }
+
+    for (let _ = 0; _ < params.CACHE_ROUNDS; _++) {
+      for (i = 0; i < n; i++) {
+        const v = o[i].readUInt32LE(0) % n
+        o[i] = keccak(xor(o[(i - 1 + n) % n], o[v]), 512)
+      }
+    }
+
+    this.cache = o
+    return this.cache
+  }
+
+  calcDatasetItem(i: number): Buffer {
+    const n = this.cache.length
+    const r = Math.floor(params.HASH_BYTES / params.WORD_BYTES)
+    let mix = Buffer.from(this.cache[i % n])
+    mix.writeInt32LE(mix.readUInt32LE(0) ^ i, 0)
+    mix = keccak(mix, 512)
+    for (let j = 0; j < params.DATASET_PARENTS; j++) {
+      const cacheIndex = fnv(i ^ j, mix.readUInt32LE((j % r) * 4))
+      mix = fnvBuffer(mix, this.cache[cacheIndex % n])
+    }
+    return keccak(mix, 512)
+  }
+
+  run(val: Buffer, nonce: Buffer, fullSize?: number) {
+    if (!fullSize && this.fullSize) {
+      fullSize = this.fullSize
+    }
+    if (!fullSize) {
+      throw new Error('fullSize needed')
+    }
+    const n = Math.floor(fullSize / params.HASH_BYTES)
+    const w = Math.floor(params.MIX_BYTES / params.WORD_BYTES)
+    const s = keccak(Buffer.concat([val, bufReverse(nonce)]), 512)
+    const mixhashes = Math.floor(params.MIX_BYTES / params.HASH_BYTES)
+    let mix = Buffer.concat(Array(mixhashes).fill(s))
+
+    let i
+    for (i = 0; i < params.ACCESSES; i++) {
+      const p =
+        (fnv(i ^ s.readUInt32LE(0), mix.readUInt32LE((i % w) * 4)) %
+          Math.floor(n / mixhashes)) *
+        mixhashes
+      let newdata = []
+      for (let j = 0; j < mixhashes; j++) {
+        newdata.push(this.calcDatasetItem(p + j))
+      }
+      mix = fnvBuffer(mix, Buffer.concat(newdata))
+    }
+
+    const cmix = Buffer.alloc(mix.length / 4)
+    for (i = 0; i < mix.length / 4; i = i + 4) {
+      const a = fnv(mix.readUInt32LE(i * 4), mix.readUInt32LE((i + 1) * 4))
+      const b = fnv(a, mix.readUInt32LE((i + 2) * 4))
+      const c = fnv(b, mix.readUInt32LE((i + 3) * 4))
+      cmix.writeUInt32LE(c, i)
+    }
```

Comment:
Can't really judge/follow why this has moved into the for loop, can you give a short explanation?

Reply:
Sorry, misread the brackets from the following js code: 🙂

```javascript
for (i = 0; i < mix.length / 4; i = i + 4) {
    var a = ethHashUtil.fnv(
      mix.readUInt32LE(i * 4),
      mix.readUInt32LE((i + 1) * 4)
    )
    var b = ethHashUtil.fnv(a, mix.readUInt32LE((i + 2) * 4))
    var c = ethHashUtil.fnv(b, mix.readUInt32LE((i + 3) * 4))
    cmix.writeUInt32LE(c, i)
  }
```

So no issue here.